### PR TITLE
Use `PUBLISHED DESC` as default return order, add sortBy filter list support.

### PR DESCRIPTION
**Pull request by Shinsina**
Screen Shot 2022-01-18 at 11 58 02 AM
Screen Shot 2022-01-18 at 11 58 13 AM
Screen Shot 2022-01-18 at 11 58 21 AM

### CODE REVIEW

**Comment by Shinsina** [comment]
[Add sort "by date" option to search results](https://trello.com/c/ivtGKIi1/131-add-sort-by-date-option-to-search-results)

**Review by solocommand — DISMISSED** [review]
@Shinsina I think I see what you're doing here, but a couple UX changes:

`Sort by` should always be present, not only show up when using a search term. It shouldn't be referenced as a filter in the code (and IMO should be above the filters, not below).

All available sorting options should be visible in the sort by section, and unselectable items should be disabled (such as when a search term is not present).

The query loader shouldn't care about the search term being present, revert the default value and sort default changes -- this should be handled by the higher level component. Ideally this is just a function of the sort by list.

Just pass `SCORE` as the sortField when selecting relevance and this should work as-is.

**Comment by Shinsina** [comment]
New comment to post various screenshots for the most recent set of updates to this PR:

Desktop, with searchQuery and sort applied:
Screen Shot 2022-01-17 at 1 34 49 PM

Desktop, search landing page: 
Screen Shot 2022-01-17 at 1 35 22 PM

Desktop, with existing filters applied only:
Screen Shot 2022-01-17 at 1 35 33 PM

Desktop, with existing filters AND searchQuery: 
Screen Shot 2022-01-17 at 1 35 44 PM

Desktop, with existing filters, searchQuery AND sort applied: 
Screen Shot 2022-01-17 at 1 35 57 PM

Mobile screenshots of previous:

Screen Shot 2022-01-17 at 1 42 10 PM
Screen Shot 2022-01-17 at 1 43 45 PM
Screen Shot 2022-01-17 at 1 50 17 PM

File: packages/marko-web-search/browser/sort-select.vue [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+</template>
+
+<script>
+export default {
+  props: {
+    className: {
+      type: String,
+      default: 'marko-web-search-sort-select',
+    },
+    nodes: {
```
Comment (zarathustra323) [inline packages/marko-web-search/browser/sort-select.vue]: Rename `nodes` to `options`

File: packages/marko-web-search/browser/sort-select.vue [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      :value="item.id"
+    >
+      Sort: {{ item.label }}
+    </option>
+  </select>
+</template>
+
+<script>
+export default {
+  props: {
```
Comment (zarathustra323) [inline packages/marko-web-search/browser/sort-select.vue:16]: Add `selectedId` prop and pass from marko component typed as a `String` with a default of `published_desc`

File: packages/marko-web-search/browser/sort-select.vue [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<template>
+  <select :class="className" @change="onChange">
+    <option
+      v-for="item in nodes"
+      :key="item.id"
+      :selected="item.isSelected"
```
Comment (zarathustra323) [inline packages/marko-web-search/browser/sort-select.vue]: `:selected="item.id === selectedId"`

File: packages/marko-web-search/browser/sort-select.vue [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      type: String,
+      default: 'marko-web-search-sort-select',
+    },
+    nodes: {
+      type: Array,
+      required: true,
+    },
+  },
+  methods: {
+    onChange(e) {
```
Comment (zarathustra323) [inline packages/marko-web-search/browser/sort-select.vue]: Change from `e` to `event`

File: packages/marko-web-search/components/filters/selected/index.marko [hunk 5]
```diff
@@ -3,8 +3,8 @@ import { get, getAsArray, getAsObject } from "@parameter1/base-cms-object-path";
 
 $ const { $markoWebSearch: search } = out.global;
 $ const blockName = "marko-web-search-selected-filters";
-$ const { contentTypes, assignedToWebsiteSectionIds } = search.input;
```
Comment (zarathustra323) [inline packages/marko-web-search/components/filters/selected/index.marko]: Discard all changes here

File: packages/marko-web-search/components/sort-by/index.marko [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { get, getAsArray, getAsObject } from "@parameter1/base-cms-object-path";
+import { dasherize } from "@parameter1/base-cms-inflector";
+import defaultValue from "@parameter1/base-cms-marko-core/utils/default-value";
+
+$ const { $markoWebSearch: search } = out.global;
+
+$ const blockName = "marko-web-search-sort-by";
+$ const items = search.config.sortBy;
```
Comment (zarathustra323) [inline packages/marko-web-search/components/sort-by/index.marko]: Rename `items` to `options` and hard set here.
```js
const options = [
  { id: 'PUBLISHED', label 'Published' },
  { id: 'SCORE', label 'Relevancy' },
]
```

File: packages/marko-web-search/components/sort-by/index.marko [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+<if(items.length)>
```
Comment (zarathustra323) [inline packages/marko-web-search/components/sort-by/index.marko]: This won't be needed now

File: packages/marko-web-search/config/index.js [hunk 8]
```diff
@@ -24,12 +24,15 @@ class MarkoWebSearchConfig {
    *                                                      Defaults to none. Should be an array of
    *                                                      section IDs, e.g. [123, 321]
    *
+   * @param {object[]} [params.sortBy] A list of filters for search results,
```
Comment (zarathustra323) [inline packages/marko-web-search/config/index.js]: Discard changes

File: packages/marko-web-search/config/query-params.js [hunk 9]
```diff
@@ -66,6 +66,12 @@ class MarkoWebSearchQueryParamConfig {
         defaultValue: () => [],
         toInput: toIntArrayInput,
         fromInput: fromArrayInput,
+      })
+      .add('sortBy', {
```
Comment (zarathustra323) [inline packages/marko-web-search/config/query-params.js]: Change `sortBy` to `sortField`

File: packages/marko-web-search/config/query-params.js [hunk 10]
```diff
@@ -66,6 +66,12 @@ class MarkoWebSearchQueryParamConfig {
...
+        type: Array,
+        defaultValue: () => [],
+        toInput: toStringArrayInput,
+        fromInput: fromArrayInput,
       });
```
Comment (zarathustra323) [inline packages/marko-web-search/config/query-params.js:81]: At top of file, add: `const sortFieldSet = new Set('PUBLISHED', 'SCORE')`;

Add default: 
`defaultValue: 'PUBLISHED'`

Add validator:

```js
validator: v => sortFieldSet.has(v),
```

File: packages/marko-web-search/config/query-params.js [hunk 11]
```diff
@@ -66,6 +66,12 @@ class MarkoWebSearchQueryParamConfig {
         defaultValue: () => [],
         toInput: toIntArrayInput,
         fromInput: fromArrayInput,
+      })
+      .add('sortBy', {
+        type: Array,
+        defaultValue: () => [],
+        toInput: toStringArrayInput,
```
Comment (zarathustra323) [inline packages/marko-web-search/config/query-params.js]: Remove `toInput` and `fromInput`

File: packages/marko-web-search/config/query-params.js [hunk 12]
```diff
@@ -66,6 +66,12 @@ class MarkoWebSearchQueryParamConfig {
         defaultValue: () => [],
         toInput: toIntArrayInput,
         fromInput: fromArrayInput,
+      })
```
Comment (zarathustra323) [inline packages/marko-web-search/config/query-params.js:71]: Add: 
```js
.add('sortOrder', {
  type: String,
  defaultValue: 'DESC',
  validator: v => (v === 'DESC' || v === 'ASC'),
})
```

File: packages/marko-web-search/components/sort-by/index.marko [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    $ const nodes = search.input.sortBy.length ? items.map((item) => {
+      const isSelected = item.id === search.input.sortBy[0] ? true : false;
+      return { ...item, isSelected };
+    }) : items.map((item, index) => {
+      const isSelected = index ? false : true;
+      return { ...item, isSelected }
+    });
+    $ const { searchQuery } = search.input;
+    <if(searchQuery)>
+      <marko-web-browser-component name="MarkoWebSearchSortSelect" props={ nodes, searchQuery } />
```
Comment (zarathustra323) [inline packages/marko-web-search/components/sort-by/index.marko]: Pass `sort-field=search.input.sortField` from website code, then `props={ options, selectedId: input.sortField }`

File: packages/marko-web-search/components/sort-by/index.marko [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import defaultValue from "@parameter1/base-cms-marko-core/utils/default-value";
+
+$ const { $markoWebSearch: search } = out.global;
+
+$ const blockName = "marko-web-search-sort-by";
+$ const items = search.config.sortBy;
+
+<if(items.length)>
+  <marko-web-block name=blockName modifiers=input.modifiers>
+    $ const nodes = search.input.sortBy.length ? items.map((item) => {
```
Comment (zarathustra323) [inline packages/marko-web-search/components/sort-by/index.marko]: The node mapping can go away

File: packages/marko-web-search/components/sort-by/index.marko [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+      const isSelected = index ? false : true;
+      return { ...item, isSelected }
+    });
+    $ const { searchQuery } = search.input;
+    <if(searchQuery)>
+      <marko-web-browser-component name="MarkoWebSearchSortSelect" props={ nodes, searchQuery } />
+    </if>
+    <else>
+      <script>
+        const urlSearchParams = new URLSearchParams(window.location.search);
```
Comment (zarathustra323) [inline packages/marko-web-search/components/sort-by/index.marko]: This can go away

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/marko-web-search/browser/sort-select.vue [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    },
+    selectedId: {
+      type: String,
+      default: 'PUBLISHED',
+    },
+  },
+  methods: {
+    onChange(event) {
+      const { value } = event.target;
+      const urlSearchParams = new URLSearchParams(window.location.search);
```
Comment (zarathustra323) [inline packages/marko-web-search/browser/sort-select.vue]:
```js
const url = new URL(window.location);
const params = new URLSearchParams(window.location.search);
params.set('sortField', value);
url.search = `${params}`;
window.location.href = `${url}`
```

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/marko-web-search/browser/sort-select.vue [hunk 17]
```diff
@@ -0,0 +1,41 @@
+<template>
+  <select :class="className" @change="onChange">
```
Comment (zarathustra323) [inline packages/marko-web-search/browser/sort-select.vue]: Change to `<select class="custom-select" @change="onChange">` and then remove the `className` from the props list.

File: packages/marko-web-search/components/query.marko [hunk 18]
```diff
@@ -20,7 +20,7 @@ $ const params = {
     : { assignedToWebsiteSiteIds: [config.website("id")] }
   ),
   queryFragment: input.queryFragment,
-  opSuffix: input.opSuffix,
```
Comment (zarathustra323) [inline packages/marko-web-search/components/query.marko]: Restore the trailing comma

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Review by zarathustra323 — APPROVED** [review]

**Pull request merged**